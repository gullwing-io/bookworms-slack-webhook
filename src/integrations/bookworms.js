import { loadBookmarks, generateBookmarks } from "bookworms";

let bookmarks = "";

const init = async (path) => {
  const { body } = await loadBookmarks.fetchBookmarkConfig(path);
  bookmarks = body;
};

// generating the markdown only for the top level folder requested in the slash command
const generateBookmarkMarkdown = (folder) => {
  const requestedFolder = findBookMarkFolder(folder, bookmarks.folders);
  if (requestedFolder) {
    const selectedBookmarks = {
      folders: [requestedFolder],
    };
    const [browser, readme] =
      generateBookmarks.generateImportBookmarkMarkup(selectedBookmarks);
    // taking the readme as this is markdown
    return readme.body;
  } else {
    return `Sorry ${folder} could not be found, use \`/bookmarks all\` for a list of available bookmarks    `;
  }
};

// used for showing a dynamic list of available top level folders
const listOfBookmarksMarkDown = () => {
  const commands = bookmarks.folders.map((folder) => {
    return `\`/bookmarks ${folder.label.toLowerCase()}\``;
  });
  // adding in new line for formatting in Slack response
  const commandsMarkdown = `
${commands.join("\r\n")}
    `;
  return `Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands
    ${commandsMarkdown}
    `;
};

// returning the markdown of the requested top level folder
const findBookMarkFolder = (name, folders) => {
  return folders.find((folder) => {
    if (folder.label.toLowerCase() === name) {
      return folder;
    }
  });
};

const getBookmarks = (folder) => generateBookmarkMarkdown(folder);

export { init, getBookmarks, listOfBookmarksMarkDown, findBookMarkFolder };
