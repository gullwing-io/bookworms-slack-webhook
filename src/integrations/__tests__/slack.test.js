import { jest } from "@jest/globals";
import { init } from "../../integrations/bookworms.js";
import {
  sendBookMarks,
  sendBookmarkCommands,
  headerForSlackMessage,
} from "../slack.js";

// I rather not test against real data but mocks with native imports are causing me issues
await init(
  "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
);

describe("Slack", () => {
  describe("headerForSlackMessage", () => {
    test("should return markdown for slack tite", () => {
      expect(headerForSlackMessage("sample tite")).toEqual([
        { text: { text: "sample tite", type: "plain_text" }, type: "header" },
        { type: "divider" },
      ]);
    });
  });
  describe("sendBookmarkCommands", () => {
    test("should return markdown for all commands", () => {
      expect(sendBookmarkCommands()).toMatchSnapshot();
    });
  });
  describe("sendBookMarks", () => {
    test("should return markdown unfound bookmarks", () => {
      expect(sendBookMarks()).toMatchSnapshot();
    });
    test("should return markdown found bookmarks", () => {
      expect(sendBookMarks("folder 1")).toMatchSnapshot();
    });
  });
});
