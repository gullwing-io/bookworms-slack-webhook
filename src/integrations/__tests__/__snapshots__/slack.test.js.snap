// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slack sendBookMarks should return markdown found bookmarks 1`] = `
Object {
  "blocks": Array [
    Object {
      "text": Object {
        "text": "Bookmarks",
        "type": "plain_text",
      },
      "type": "header",
    },
    Object {
      "type": "divider",
    },
    Object {
      "text": Object {
        "text": "*folder 1*

This is to describe the folder structure

*sub folder 1*

This is to describe the sub folder structure

•   <https://www.mywebsite.com|sample url 1> - this is used to describe the bookmark

​_These bookmarks were last updated on 26/03/2022, 18:09:37 using <https://github.com/thearegee/bookworms|Bookworms>_​
",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;

exports[`Slack sendBookMarks should return markdown unfound bookmarks 1`] = `
Object {
  "blocks": Array [
    Object {
      "text": Object {
        "text": "Bookmarks",
        "type": "plain_text",
      },
      "type": "header",
    },
    Object {
      "type": "divider",
    },
    Object {
      "text": Object {
        "text": "Sorry undefined could not be found, use \`/bookmarks all\` for a list of available bookmarks
",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;

exports[`Slack sendBookmarkCommands should return markdown for all commands 1`] = `
Object {
  "blocks": Array [
    Object {
      "text": Object {
        "text": "Bookmarks",
        "type": "plain_text",
      },
      "type": "header",
    },
    Object {
      "type": "divider",
    },
    Object {
      "text": Object {
        "text": "Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands

\`/bookmarks folder 1\`
\`/bookmarks folder 2\`
",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;
